pkg/networkservice/common/timeout/server.go

```diff
@@ -21,80 +21,145 @@ import (
+		if err != nil {
+			return
 		}
+
+		var timer *timer
+		timer, err = t.createTimer(ctx, conn)
+		if err != nil {
+			if _, closeErr := next.Server(ctx).Close(ctx, conn); closeErr != nil {
+				err = errors.Wrapf(err, "error attempting to close failed connection %v: %+v", connID, closeErr)
+			}
+			return
+		}
+
+		t.connections.Store(connID, timer)
 	})
-	return next.Server(ctx).Close(ctx, conn)
+
+	return conn, err
 }
 
-func (t *timeoutServer) createTimer(ctx context.Context, request *networkservice.NetworkServiceRequest) (*time.Timer, error) {
-	request = request.Clone()
-	expireTime, err := ptypes.Timestamp(request.GetConnection().GetPath().GetPathSegments()[request.GetConnection().GetPath().GetIndex()-1].GetExpires())
+func (t *timeoutServer) createTimer(ctx context.Context, conn *networkservice.Connection) (*timer, error) {
+	logEntry := log.Entry(ctx).WithField("timeoutServer", "createTimer")
+
+	expireTime, err := ptypes.Timestamp(conn.GetPath().GetPathSegments()[conn.GetPath().GetIndex()-1].GetExpires())
 	if err != nil {
 		return nil, err
 	}
-	duration := time.Until(expireTime)
-	conn := request.GetConnection().Clone()
-	return time.AfterFunc(duration, func() {
-		newCtx := extend.WithValuesFromContext(context.Background(), ctx)
-		if _, err := (*t.onTimeout).Close(newCtx, conn); err != nil {
-			trace.Log(newCtx).Errorf("Error attempting to close timed out connection: %s: %+v", request.GetConnection().GetId(), err)
+
+	conn = conn.Clone()
+	timerCtx := extend.WithValuesFromContext(context.Background(), t.ctx)
+
+	timer := &timer{
+		stopCh: make(chan struct{}),
+	}
+	timer.timer = time.AfterFunc(time.Until(expireTime), func() {
+		executor, ok := t.executors.Load(conn.GetId())
+		if !ok {
+			return
 		}
-	}), nil
+
+		executor.AsyncExec(func() {
+			select {
+			case <-timer.stopCh:
+				logEntry.Warnf("timer has been already stopped: %v", conn.GetId())
+			default:
+				if err := t.close(timerCtx, conn, next.Server(ctx)); err != nil {
+					logEntry.Errorf("failed to close timed out connection: %v %+v", conn.GetId(), err)
+				}
+			}
+		})
+	})
+
+	return timer, nil
+}
+
+func (t *timeoutServer) Close(ctx context.Context, conn *networkservice.Connection) (_ *empty.Empty, err error) {
+	logEntry := log.Entry(ctx).WithField("timeoutServer", "Close")
+
+	executor, ok := t.executors.Load(conn.GetId())
+	if !ok {
+		logEntry.Warnf("connection has been already closed: %v", conn.GetId())
+		return &empty.Empty{}, nil
+	}
+
+	<-executor.AsyncExec(func() {
+		err = t.close(ctx, conn, next.Server(ctx))
+	})
+
+	return &empty.Empty{}, err
+}
+
+func (t *timeoutServer) close(ctx context.Context, conn *networkservice.Connection, nextServer networkservice.NetworkServiceServer) error {
+	logEntry := log.Entry(ctx).WithField("timeoutServer", "close")
+
+	timer, ok := t.connections.Load(conn.GetId())
+	if !ok {
+		logEntry.Warnf("connection has been already closed: %v", conn.GetId())
+		return nil
+	}
+
+	timer.timer.Stop()
```

Comment:
Should we in case if the timer is not stopped wait for done?

Reply:
All events are executed inside `executor.AsyncExec()` - we either have already waited for the event to be done or we are running before that event and it should be stopped with `timer.Stop()` or `close(closeCh)`.